// @plasmicpkgs/react-aria/skinny/registerSliderTrack.cjs.js
'use strict';

var React = require('react');
var reactAriaComponents = require('react-aria-components');
var flattenChildren = require('react-keyed-flatten-children');
var common = require('./common-b3b54c72.cjs.js');
var contexts = require('./contexts-6d0cb2b1.cjs.js');
var utils = require('./utils-7d000fa4.cjs.js');
var variantUtils = require('./variant-utils-0ad70db8.cjs.js');
require('@plasmicapp/host');
require('@plasmicapp/host/registerComponent');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var flattenChildren__default = /*#__PURE__*/_interopDefault(flattenChildren);

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const SLIDER_TRACK_VARIANTS = ["hovered"];
const { variants, withObservedValues } = variantUtils.pickAriaComponentVariants(
  SLIDER_TRACK_VARIANTS
);
function isMultiValueGuard(value) {
  return Array.isArray(value) && value.length > 1;
}
function BaseSliderTrack(props) {
  const context = React__default.default.useContext(contexts.PlasmicSliderContext);
  const isStandalone = !context;
  const _a = props, { children, progressBar, plasmicUpdateVariant } = _a, rest = __objRest(_a, ["children", "progressBar", "plasmicUpdateVariant"]);
  const thumbsLength = context && isMultiValueGuard(context.value) ? context.value.length : 1;
  const isMultiValue = thumbsLength > 1;
  const { minIndex, maxIndex } = React.useMemo(() => {
    if (thumbsLength <= 1) {
      return { minIndex: 0, maxIndex: 0 };
    }
    return { minIndex: 0, maxIndex: thumbsLength - 1 };
  }, [thumbsLength]);
  const thumbs = React.useMemo(() => {
    const thumbNodes = flattenChildren__default.default(children);
    if (!thumbNodes || thumbNodes.length === 0 || !utils.isDefined(context == null ? void 0 : context.value)) {
      return [];
    }
    const values = utils.isDefined(context) ? Array.isArray(context.value) ? context.value : [context.value] : [];
    const lastThumb = thumbNodes[thumbNodes.length - 1];
    return values.map((v, i) => {
      const currentThumb = thumbNodes[i];
      if (i >= thumbNodes.length) {
        if (React__default.default.isValidElement(lastThumb)) {
          return React__default.default.cloneElement(lastThumb, {
            index: i
          });
        }
      }
      if (!React__default.default.isValidElement(currentThumb)) {
        return null;
      }
      return React__default.default.cloneElement(currentThumb, {
        index: i
      });
    });
  }, [children, context == null ? void 0 : context.value]);
  const track = /* @__PURE__ */ React__default.default.createElement(reactAriaComponents.SliderTrack, __spreadValues({ style: { position: "relative" } }, rest), ({ state, isHovered }) => /* @__PURE__ */ React__default.default.createElement(React__default.default.Fragment, null, withObservedValues(
    /* @__PURE__ */ React__default.default.createElement(React__default.default.Fragment, null, /* @__PURE__ */ React__default.default.createElement(
      "div",
      {
        style: {
          width: `${(!isMultiValue ? state.getThumbPercent(minIndex) : state.getThumbPercent(maxIndex) - state.getThumbPercent(minIndex)) * 100}%`,
          height: "100%",
          position: "absolute",
          top: 0,
          left: !isMultiValue ? 0 : state.getThumbPercent(minIndex) * 100 + "%"
        }
      },
      progressBar
    ), thumbs),
    {
      hovered: isHovered
    },
    plasmicUpdateVariant
  )));
  if (isStandalone) {
    return /* @__PURE__ */ React__default.default.createElement(reactAriaComponents.Slider, { style: __spreadValues({ height: "100%", width: "100%" }, common.COMMON_STYLES) }, track);
  }
  return track;
}
const SLIDER_TRACK_COMPONENT_NAME = utils.makeComponentName("sliderTrack");
function registerSliderTrack(sliderThumbMeta, loader, overrides) {
  return utils.registerComponentHelper(
    loader,
    BaseSliderTrack,
    {
      name: SLIDER_TRACK_COMPONENT_NAME,
      displayName: "Aria Slider Track",
      importPath: "@plasmicpkgs/react-aria/skinny/registerSliderTrack",
      importName: "BaseSliderTrack",
      variants,
      defaultStyles: {
        width: "stretch",
        backgroundColor: "#aaa",
        position: "relative",
        height: "10px",
        padding: 0
      },
      props: {
        children: {
          type: "slot",
          /**
           * NOTE: We don't merge with parent here, because we want to allow the user to select the thumbs without having to first select the slider track.
           * Also, there can be more than one thumbs (e.g. in a range slider), but `mergeWithParent` only shows prop controls of the slot content if there is only one direct descendant of the slot.
           * */
          // mergeWithParent: true,
          displayName: "Thumbs",
          description: "The thumbs of the slider. For range slider, you can add more than one thumb.",
          allowedComponents: [sliderThumbMeta.name],
          allowRootWrapper: true,
          defaultValue: [
            {
              type: "component",
              name: sliderThumbMeta.name
            }
          ]
        },
        progressBar: {
          type: "slot",
          mergeWithParent: true,
          displayName: "Progress Bar",
          defaultValue: [
            {
              type: "box",
              styles: {
                height: "100%",
                width: "100%",
                backgroundColor: "#ffa6a6",
                padding: 0
              }
            }
          ]
        },
        onHoverStart: {
          type: "eventHandler",
          argTypes: [{ name: "event", type: "object" }]
        },
        onHoverEnd: {
          type: "eventHandler",
          argTypes: [{ name: "event", type: "object" }]
        },
        onHoverChange: {
          type: "eventHandler",
          argTypes: [{ name: "isHovering", type: "boolean" }]
        }
      }
    },
    overrides
  );
}

exports.BaseSliderTrack = BaseSliderTrack;
exports.SLIDER_TRACK_COMPONENT_NAME = SLIDER_TRACK_COMPONENT_NAME;
exports.registerSliderTrack = registerSliderTrack;
