// @plasmicpkgs/react-aria/skinny/registerDescription.cjs.js
'use strict';

var registerText = require('./registerText.cjs.js');
var utils = require('./utils-7d000fa4.cjs.js');
require('react');
require('react-aria-components');
require('./common-b3b54c72.cjs.js');
require('@plasmicapp/host');
require('@plasmicapp/host/registerComponent');

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
const DESCRIPTION_COMPONENT_NAME = utils.makeComponentName("description");
function registerDescription(loader, overrides) {
  return registerText.registerText(loader, __spreadProps(__spreadValues({}, overrides), {
    name: DESCRIPTION_COMPONENT_NAME,
    displayName: "Aria Description",
    props: {
      slot: {
        type: "string",
        hidden: () => true,
        defaultValue: "description"
      }
    }
  }));
}

exports.DESCRIPTION_COMPONENT_NAME = DESCRIPTION_COMPONENT_NAME;
exports.registerDescription = registerDescription;
