// @plasmicpkgs/react-aria/skinny/registerDialog.esm.js
import { usePlasmicCanvasContext } from '@plasmicapp/host';
import React from 'react';
import { Dialog } from 'react-aria-components';
import { C as COMMON_STYLES } from './common-ceebbaea.esm.js';
import { m as makeComponentName, r as registerComponentHelper } from './utils-5d1b4c6b.esm.js';
import '@plasmicapp/host/registerComponent';

function BaseDialog({ children, className }) {
  const canvasContext = usePlasmicCanvasContext();
  if (canvasContext) {
    return /* @__PURE__ */ React.createElement("div", { className, style: COMMON_STYLES }, children);
  } else {
    return /* @__PURE__ */ React.createElement(Dialog, { className, style: COMMON_STYLES }, children);
  }
}
const DIALOG_COMPONENT_NAME = makeComponentName("dialog");
function registerDialog(loader, overrides) {
  registerComponentHelper(
    loader,
    BaseDialog,
    {
      name: DIALOG_COMPONENT_NAME,
      displayName: "Aria Dialog",
      importPath: "@plasmicpkgs/react-aria/skinny/registerDialog",
      importName: "BaseDialog",
      props: {
        children: {
          type: "slot",
          mergeWithParent: true
        }
      },
      trapsFocus: true
    },
    overrides
  );
}

export { BaseDialog, DIALOG_COMPONENT_NAME, registerDialog };
