// @plasmicpkgs/react-aria/skinny/registerTextArea.esm.js
import React, { useCallback } from 'react';
import { mergeProps } from 'react-aria';
import { TextArea } from 'react-aria-components';
import { g as getCommonProps, C as COMMON_STYLES } from './common-ceebbaea.esm.js';
import { g as PlasmicTextFieldContext } from './contexts-5cb81c2f.esm.js';
import { r as registerComponentHelper, m as makeComponentName } from './utils-5d1b4c6b.esm.js';
import { p as pickAriaComponentVariants } from './variant-utils-4405ebb0.esm.js';
import '@plasmicapp/host';
import '@plasmicapp/host/registerComponent';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const TEXTAREA_VARIANTS = [
  "focused",
  "focusVisible",
  "hovered",
  "disabled"
];
const { variants } = pickAriaComponentVariants(TEXTAREA_VARIANTS);
const inputHelpers = {
  states: {
    value: {
      onChangeArgsToValue: (e) => {
        return e.target.value;
      }
    }
  }
};
function BaseTextArea(props) {
  const _a = props, { className, plasmicUpdateVariant, setControlContextData } = _a, rest = __objRest(_a, ["className", "plasmicUpdateVariant", "setControlContextData"]);
  const textFieldContext = React.useContext(PlasmicTextFieldContext);
  setControlContextData == null ? void 0 : setControlContextData({
    parent: textFieldContext
  });
  const classNameProp = useCallback(
    ({
      isDisabled,
      isFocusVisible,
      isFocused,
      isHovered
    }) => {
      plasmicUpdateVariant == null ? void 0 : plasmicUpdateVariant({
        disabled: isDisabled,
        focused: isFocused,
        focusVisible: isFocusVisible,
        hovered: isHovered
      });
      return className != null ? className : "";
    },
    [className, plasmicUpdateVariant]
  );
  const mergedProps = mergeProps(rest, {
    className: classNameProp
  });
  return /* @__PURE__ */ React.createElement(TextArea, __spreadProps(__spreadValues({}, mergedProps), { style: COMMON_STYLES }));
}
function registerTextArea(loader, overrides) {
  registerComponentHelper(
    loader,
    BaseTextArea,
    {
      name: makeComponentName("textarea"),
      displayName: "Aria TextArea",
      importPath: "@plasmicpkgs/react-aria/skinny/registerTextArea",
      importName: "BaseTextArea",
      variants,
      props: __spreadValues({}, getCommonProps("Text Area", [
        "name",
        "disabled",
        "readOnly",
        "autoFocus",
        "aria-label",
        "required",
        "placeholder",
        "value",
        "maxLength",
        "minLength",
        "inputMode",
        "onChange",
        "onFocus",
        "onBlur",
        "onKeyDown",
        "onKeyUp",
        "onCopy",
        "onCut",
        "onPaste",
        "onCompositionStart",
        "onCompositionEnd",
        "onCompositionUpdate",
        "onSelect",
        "onBeforeInput",
        "onInput"
      ])),
      states: {
        value: __spreadValues({
          type: "writable",
          valueProp: "value",
          onChangeProp: "onChange",
          variableType: "text"
        }, inputHelpers.states.value)
      },
      componentHelpers: {
        helpers: inputHelpers,
        importName: "inputHelpers",
        importPath: "@plasmicpkgs/react-aria/skinny/registerTextArea"
      }
    },
    overrides
  );
}

export { BaseTextArea, inputHelpers, registerTextArea };
