// @plasmicpkgs/react-aria/skinny/registerDialog.cjs.js
'use strict';

var host = require('@plasmicapp/host');
var React = require('react');
var reactAriaComponents = require('react-aria-components');
var common = require('./common-b3b54c72.cjs.js');
var utils = require('./utils-7d000fa4.cjs.js');
require('@plasmicapp/host/registerComponent');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

function BaseDialog({ children, className }) {
  const canvasContext = host.usePlasmicCanvasContext();
  if (canvasContext) {
    return /* @__PURE__ */ React__default.default.createElement("div", { className, style: common.COMMON_STYLES }, children);
  } else {
    return /* @__PURE__ */ React__default.default.createElement(reactAriaComponents.Dialog, { className, style: common.COMMON_STYLES }, children);
  }
}
const DIALOG_COMPONENT_NAME = utils.makeComponentName("dialog");
function registerDialog(loader, overrides) {
  utils.registerComponentHelper(
    loader,
    BaseDialog,
    {
      name: DIALOG_COMPONENT_NAME,
      displayName: "Aria Dialog",
      importPath: "@plasmicpkgs/react-aria/skinny/registerDialog",
      importName: "BaseDialog",
      props: {
        children: {
          type: "slot",
          mergeWithParent: true
        }
      },
      trapsFocus: true
    },
    overrides
  );
}

exports.BaseDialog = BaseDialog;
exports.DIALOG_COMPONENT_NAME = DIALOG_COMPONENT_NAME;
exports.registerDialog = registerDialog;
