// @plasmicpkgs/react-aria/skinny/registerComboBox.esm.js
import React, { useCallback, useMemo, useEffect } from 'react';
import { ComboBox, ComboBoxStateContext } from 'react-aria-components';
import { g as getCommonProps, C as COMMON_STYLES, a as arrowDown } from './common-ceebbaea.esm.js';
import { a as PlasmicPopoverTriggerContext, b as PlasmicListBoxContext, c as PlasmicInputContext } from './contexts-5cb81c2f.esm.js';
import { O as OptionsItemIdManager } from './OptionsItemIdManager-832b8cfe.esm.js';
import { BUTTON_COMPONENT_NAME } from './registerButton.esm.js';
import { INPUT_COMPONENT_NAME } from './registerInput.esm.js';
import { LABEL_COMPONENT_NAME } from './registerLabel.esm.js';
import { L as LIST_BOX_COMPONENT_NAME } from './registerListBox-40846430.esm.js';
import { POPOVER_COMPONENT_NAME } from './registerPopover.esm.js';
import { r as registerComponentHelper, m as makeComponentName, u as useAutoOpen } from './utils-5d1b4c6b.esm.js';
import { p as pickAriaComponentVariants } from './variant-utils-4405ebb0.esm.js';
import '@plasmicapp/host';
import 'react-aria';
import './registerListBoxItem.esm.js';
import './registerDescription.esm.js';
import './registerText.esm.js';
import '@react-aria/utils';
import '@plasmicapp/host/registerComponent';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const COMBOBOX_NAME = makeComponentName("combobox");
const COMBOBOX_VARIANTS = ["disabled"];
const { variants: COMBOBOX_VARIANTS_DATA } = pickAriaComponentVariants(COMBOBOX_VARIANTS);
function ComboboxAutoOpen(props) {
  var _a;
  const { open, close } = (_a = React.useContext(ComboBoxStateContext)) != null ? _a : {};
  useAutoOpen({ props, open, close });
  return null;
}
function BaseComboBox(props) {
  const _a = props, {
    children,
    setControlContextData,
    plasmicUpdateVariant,
    className,
    isOpen: _isOpen
  } = _a, rest = __objRest(_a, [
    "children",
    "setControlContextData",
    "plasmicUpdateVariant",
    "className",
    "isOpen"
  ]);
  const classNameProp = useCallback(
    ({ isDisabled }) => {
      plasmicUpdateVariant == null ? void 0 : plasmicUpdateVariant({
        disabled: isDisabled
      });
      return className != null ? className : "";
    },
    [className, plasmicUpdateVariant]
  );
  const idManager = useMemo(() => new OptionsItemIdManager(), []);
  useEffect(() => {
    idManager.subscribe((ids) => {
      setControlContextData == null ? void 0 : setControlContextData({
        itemIds: ids
      });
    });
  }, []);
  return /* @__PURE__ */ React.createElement(ComboBox, __spreadProps(__spreadValues({ className: classNameProp }, rest), { style: COMMON_STYLES }), /* @__PURE__ */ React.createElement(PlasmicPopoverTriggerContext.Provider, { value: true }, /* @__PURE__ */ React.createElement(
    PlasmicListBoxContext.Provider,
    {
      value: {
        idManager
      }
    },
    /* @__PURE__ */ React.createElement(PlasmicInputContext.Provider, { value: { isUncontrolled: true } }, /* @__PURE__ */ React.createElement(ComboboxAutoOpen, __spreadValues({}, props)), children)
  )));
}
function registerComboBox(loader) {
  registerComponentHelper(loader, BaseComboBox, {
    name: COMBOBOX_NAME,
    displayName: "Aria ComboBox",
    importPath: "@plasmicpkgs/react-aria/skinny/registerComboBox",
    importName: "BaseComboBox",
    variants: COMBOBOX_VARIANTS_DATA,
    props: __spreadProps(__spreadValues({}, getCommonProps("ComboBox", [
      "name",
      "aria-label",
      "isDisabled"
    ])), {
      selectedKey: {
        type: "choice",
        editOnly: true,
        uncontrolledProp: "defaultSelectedKey",
        displayName: "Initial selected item",
        options: (_props, ctx) => (ctx == null ? void 0 : ctx.itemIds) ? Array.from(ctx.itemIds) : [],
        // React Aria ComboBox do not support multiple selections yet
        multiSelect: false
      },
      disabledKeys: {
        type: "choice",
        displayName: "Disabled values",
        description: "The items that are disabled. These items cannot be selected, focused, or otherwise interacted with.",
        options: (_props, ctx) => (ctx == null ? void 0 : ctx.itemIds) ? Array.from(ctx.itemIds) : [],
        multiSelect: true,
        advanced: true
      },
      isOpen: {
        type: "boolean",
        defaultValue: false,
        // It doesn't make sense to make isOpen prop editable (it's controlled by user interaction and always closed by default), so we keep this prop hidden. We have listed the prop here in the meta only so we can expose a writeable state for it.
        hidden: () => true
      },
      onSelectionChange: {
        type: "eventHandler",
        argTypes: [{ name: "selectedValue", type: "string" }]
      },
      onOpenChange: {
        type: "eventHandler",
        argTypes: [{ name: "isOpen", type: "boolean" }]
      },
      children: {
        type: "slot",
        defaultValue: [
          {
            type: "vbox",
            styles: {
              justifyContent: "flex-start",
              alignItems: "flex-start",
              width: "300px",
              padding: 0
            },
            children: [
              {
                type: "component",
                name: LABEL_COMPONENT_NAME,
                props: {
                  children: {
                    type: "text",
                    value: "Label"
                  }
                }
              },
              {
                type: "hbox",
                styles: {
                  padding: 0
                },
                children: [
                  {
                    type: "component",
                    name: INPUT_COMPONENT_NAME,
                    styles: {
                      width: "100%",
                      borderRightWidth: 0
                    }
                  },
                  {
                    type: "component",
                    name: BUTTON_COMPONENT_NAME,
                    props: {
                      children: arrowDown
                    }
                  }
                ]
              },
              {
                type: "component",
                name: POPOVER_COMPONENT_NAME,
                styles: {
                  backgroundColor: "white",
                  padding: "10px",
                  overflow: "scroll",
                  width: "unset"
                },
                props: {
                  offset: 0,
                  children: [
                    {
                      type: "component",
                      name: LIST_BOX_COMPONENT_NAME,
                      props: {
                        selectionMode: "single"
                      },
                      styles: {
                        borderWidth: 0,
                        width: "stretch"
                      }
                    }
                  ]
                }
              }
            ]
          }
        ]
      }
    }),
    states: {
      selectedValue: {
        type: "writable",
        valueProp: "selectedKey",
        onChangeProp: "onSelectionChange",
        variableType: "text"
      },
      isOpen: {
        type: "writable",
        valueProp: "isOpen",
        onChangeProp: "onOpenChange",
        variableType: "boolean"
      }
    },
    trapsFocus: true
  });
}

export { BaseComboBox, registerComboBox };
