// @plasmicpkgs/react-aria/skinny/registerDialogTrigger.cjs.js
'use strict';

var host = require('@plasmicapp/host');
var React = require('react');
var reactAriaComponents = require('react-aria-components');
var contexts = require('./contexts-6d0cb2b1.cjs.js');
var registerButton = require('./registerButton.cjs.js');
var registerDialog = require('./registerDialog.cjs.js');
var registerModal = require('./registerModal.cjs.js');
var utils = require('./utils-7d000fa4.cjs.js');
require('react-aria');
require('./common-b3b54c72.cjs.js');
require('./variant-utils-0ad70db8.cjs.js');
require('@plasmicapp/host/registerComponent');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
function BaseDialogTrigger(props) {
  var _b, _c, _d;
  const _a = props, { trigger, dialog, isOpen } = _a, rest = __objRest(_a, ["trigger", "dialog", "isOpen"]);
  const { isSelected, selectedSlotName } = (_c = (_b = host.usePlasmicCanvasComponentInfo) == null ? void 0 : _b(props)) != null ? _c : {};
  const isAutoOpen = selectedSlotName !== "trigger" && isSelected;
  const mergedProps = __spreadProps(__spreadValues({}, rest), {
    isOpen: (_d = isAutoOpen || isOpen) != null ? _d : false
  });
  return /* @__PURE__ */ React__default.default.createElement(contexts.PlasmicDialogTriggerContext.Provider, { value: mergedProps }, /* @__PURE__ */ React__default.default.createElement(reactAriaComponents.DialogTrigger, __spreadValues({}, mergedProps), trigger, dialog));
}
function registerDialogTrigger(loader, overrides) {
  utils.registerComponentHelper(
    loader,
    BaseDialogTrigger,
    {
      name: utils.makeComponentName("dialogTrigger"),
      displayName: "Aria Dialog Trigger",
      importPath: "@plasmicpkgs/react-aria/skinny/registerDialogTrigger",
      importName: "BaseDialogTrigger",
      styleSections: false,
      props: {
        trigger: {
          type: "slot",
          mergeWithParent: true,
          defaultValue: {
            type: "component",
            name: registerButton.BUTTON_COMPONENT_NAME,
            props: {
              children: {
                type: "text",
                value: "Open Dialog"
              }
            }
          }
        },
        dialog: {
          type: "slot",
          mergeWithParent: true,
          defaultValue: [
            {
              type: "component",
              name: registerModal.MODAL_COMPONENT_NAME,
              props: {
                children: {
                  type: "component",
                  name: registerDialog.DIALOG_COMPONENT_NAME,
                  props: {
                    children: registerModal.MODAL_DEFAULT_SLOT_CONTENT
                  }
                }
              }
            }
          ]
        },
        isOpen: {
          type: "boolean",
          defaultValueHint: false,
          editOnly: true,
          uncontrolledProp: "defaultOpen"
        },
        onOpenChange: {
          type: "eventHandler",
          argTypes: [{ name: "isOpen", type: "boolean" }]
        }
      },
      states: {
        isOpen: {
          type: "writable",
          valueProp: "isOpen",
          onChangeProp: "onOpenChange",
          variableType: "boolean"
        }
      },
      trapsFocus: true
    },
    overrides
  );
}

exports.BaseDialogTrigger = BaseDialogTrigger;
exports.registerDialogTrigger = registerDialogTrigger;
