// @plasmicpkgs/react-aria/skinny/registerInput.esm.js
import React, { useCallback } from 'react';
import { mergeProps } from 'react-aria';
import { Input } from 'react-aria-components';
import { r as resolveAutoComplete, g as getCommonProps, C as COMMON_STYLES } from './common-ceebbaea.esm.js';
import { g as PlasmicTextFieldContext, c as PlasmicInputContext } from './contexts-5cb81c2f.esm.js';
import { m as makeComponentName, r as registerComponentHelper } from './utils-5d1b4c6b.esm.js';
import { p as pickAriaComponentVariants } from './variant-utils-4405ebb0.esm.js';
import '@plasmicapp/host';
import '@plasmicapp/host/registerComponent';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const INPUT_VARIANTS = [
  "focused",
  "focusVisible",
  "hovered",
  "disabled"
];
const { variants } = pickAriaComponentVariants(INPUT_VARIANTS);
const inputHelpers = {
  states: {
    value: {
      onChangeArgsToValue: (e) => {
        return e.target.value;
      }
    }
  }
};
function BaseInput(props) {
  const _a = props, {
    plasmicUpdateVariant,
    setControlContextData,
    autoComplete,
    value,
    className
  } = _a, rest = __objRest(_a, [
    "plasmicUpdateVariant",
    "setControlContextData",
    "autoComplete",
    "value",
    "className"
  ]);
  const textFieldContext = React.useContext(PlasmicTextFieldContext);
  const context = React.useContext(PlasmicInputContext);
  setControlContextData == null ? void 0 : setControlContextData({
    parent: textFieldContext
  });
  const classNameProp = useCallback(
    ({
      isDisabled,
      isFocusVisible,
      isFocused,
      isHovered
    }) => {
      plasmicUpdateVariant == null ? void 0 : plasmicUpdateVariant({
        disabled: isDisabled,
        focused: isFocused,
        focusVisible: isFocusVisible,
        hovered: isHovered
      });
      return className != null ? className : "";
    },
    [className, plasmicUpdateVariant]
  );
  const mergedProps = mergeProps(rest, {
    value: (context == null ? void 0 : context.isUncontrolled) ? void 0 : value,
    autoComplete: resolveAutoComplete(autoComplete),
    className: classNameProp
  });
  return /* @__PURE__ */ React.createElement(Input, __spreadProps(__spreadValues({}, mergedProps), { style: COMMON_STYLES }));
}
const INPUT_COMPONENT_NAME = makeComponentName("input");
function registerInput(loader, overrides) {
  registerComponentHelper(
    loader,
    BaseInput,
    {
      name: INPUT_COMPONENT_NAME,
      displayName: "Aria Input",
      importPath: "@plasmicpkgs/react-aria/skinny/registerInput",
      importName: "BaseInput",
      variants,
      defaultStyles: {
        width: "300px",
        borderWidth: "1px",
        borderStyle: "solid",
        borderColor: "black",
        padding: "4px 10px"
      },
      props: __spreadValues({}, getCommonProps("Input", [
        "name",
        "disabled",
        "readOnly",
        "autoFocus",
        "aria-label",
        "required",
        "placeholder",
        "value",
        "maxLength",
        "minLength",
        "pattern",
        "type",
        "inputMode",
        "autoComplete",
        "onChange",
        "onFocus",
        "onBlur",
        "onKeyDown",
        "onKeyUp",
        "onCopy",
        "onCut",
        "onPaste",
        "onCompositionStart",
        "onCompositionEnd",
        "onCompositionUpdate",
        "onSelect",
        "onBeforeInput",
        "onInput"
      ])),
      states: {
        value: __spreadValues({
          type: "writable",
          valueProp: "value",
          onChangeProp: "onChange",
          variableType: "text"
        }, inputHelpers.states.value)
      },
      componentHelpers: {
        helpers: inputHelpers,
        importName: "inputHelpers",
        importPath: "@plasmicpkgs/react-aria/skinny/registerInput"
      }
    },
    overrides
  );
}

export { BaseInput, INPUT_COMPONENT_NAME, inputHelpers, registerInput };
