// @plasmicpkgs/react-aria/skinny/registerSliderOutput.esm.js
import React from 'react';
import { SliderOutput } from 'react-aria-components';
import { C as COMMON_STYLES } from './common-ceebbaea.esm.js';
import { m as makeComponentName, r as registerComponentHelper } from './utils-5d1b4c6b.esm.js';
import { p as pickAriaComponentVariants } from './variant-utils-4405ebb0.esm.js';
import '@plasmicapp/host';
import '@plasmicapp/host/registerComponent';

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const SLIDER_OUTPUT_VARIANTS = ["disabled"];
const { variants, withObservedValues } = pickAriaComponentVariants(
  SLIDER_OUTPUT_VARIANTS
);
function BaseSliderOutput(props) {
  const _a = props, { plasmicUpdateVariant, children } = _a, rest = __objRest(_a, ["plasmicUpdateVariant", "children"]);
  return /* @__PURE__ */ React.createElement(SliderOutput, __spreadProps(__spreadValues({}, rest), { style: COMMON_STYLES }), ({ isDisabled }) => withObservedValues(
    children,
    {
      disabled: isDisabled
    },
    plasmicUpdateVariant
  ));
}
const SLIDER_OUTPUT_COMPONENT_NAME = makeComponentName("sliderOutput");
function registerSliderOutput(loader, overrides) {
  return registerComponentHelper(
    loader,
    BaseSliderOutput,
    {
      name: SLIDER_OUTPUT_COMPONENT_NAME,
      displayName: "Aria Slider Output",
      importPath: "@plasmicpkgs/react-aria/skinny/registerSliderOutput",
      importName: "BaseSliderOutput",
      variants,
      props: {
        children: {
          mergeWithParent: true,
          type: "slot"
        }
      },
      trapsFocus: true
    },
    overrides
  );
}

export { BaseSliderOutput, SLIDER_OUTPUT_COMPONENT_NAME, registerSliderOutput };
