// @plasmicpkgs/react-aria/skinny/registerTextField.cjs.js
'use strict';

var React = require('react');
var reactAriaComponents = require('react-aria-components');
var common = require('./common-b3b54c72.cjs.js');
var contexts = require('./contexts-6d0cb2b1.cjs.js');
var registerDescription = require('./registerDescription.cjs.js');
var registerInput = require('./registerInput.cjs.js');
var registerLabel = require('./registerLabel.cjs.js');
var utils = require('./utils-7d000fa4.cjs.js');
var variantUtils = require('./variant-utils-0ad70db8.cjs.js');
require('./registerText.cjs.js');
require('react-aria');
require('@plasmicapp/host');
require('@plasmicapp/host/registerComponent');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
const TEXT_FIELD_VARIANTS = ["disabled", "readonly"];
const { variants, withObservedValues } = variantUtils.pickAriaComponentVariants(TEXT_FIELD_VARIANTS);
function BaseTextField(props) {
  const _a = props, { children, plasmicUpdateVariant, autoComplete } = _a, rest = __objRest(_a, ["children", "plasmicUpdateVariant", "autoComplete"]);
  return /* @__PURE__ */ React__default.default.createElement(contexts.PlasmicTextFieldContext.Provider, { value: props }, /* @__PURE__ */ React__default.default.createElement(
    reactAriaComponents.TextField,
    __spreadProps(__spreadValues({
      autoComplete: common.resolveAutoComplete(autoComplete)
    }, rest), {
      style: common.COMMON_STYLES
    }),
    ({ isDisabled, isReadOnly }) => withObservedValues(
      children,
      {
        disabled: isDisabled,
        readonly: isReadOnly
      },
      plasmicUpdateVariant
    )
  ));
}
const TEXT_FIELD_COMPONENT_NAME = utils.makeComponentName("textField");
function registerTextField(loader, overrides) {
  utils.registerComponentHelper(
    loader,
    BaseTextField,
    {
      name: TEXT_FIELD_COMPONENT_NAME,
      displayName: "Aria TextField",
      importPath: "@plasmicpkgs/react-aria/skinny/registerTextField",
      importName: "BaseTextField",
      variants,
      // TODO: Support for validate prop
      props: __spreadProps(__spreadValues({}, common.getCommonProps("Text Field", [
        "name",
        "isDisabled",
        "isReadOnly",
        "autoFocus",
        "aria-label",
        "isRequired",
        "value",
        "maxLength",
        "minLength",
        "pattern",
        "type",
        "inputMode",
        "validationBehavior",
        "autoComplete",
        "onChange",
        "onFocus",
        "onBlur",
        "onFocusChange",
        "onKeyDown",
        "onKeyUp",
        "onCopy",
        "onCut",
        "onPaste",
        "onCompositionStart",
        "onCompositionEnd",
        "onCompositionUpdate",
        "onSelect",
        "onBeforeInput",
        "onInput"
      ])), {
        children: {
          type: "slot",
          mergeWithParent: true,
          defaultValue: {
            type: "vbox",
            styles: {
              justifyContent: "flex-start",
              alignItems: "flex-start",
              width: "300px",
              gap: "5px",
              padding: 0
            },
            children: [
              {
                type: "component",
                name: registerLabel.LABEL_COMPONENT_NAME,
                props: {
                  children: {
                    type: "text",
                    value: "Label"
                  }
                }
              },
              {
                type: "component",
                name: registerInput.INPUT_COMPONENT_NAME,
                styles: {
                  width: "100%"
                }
              },
              {
                type: "component",
                name: registerDescription.DESCRIPTION_COMPONENT_NAME,
                props: {
                  children: {
                    type: "text",
                    value: "Type something..."
                  }
                }
              }
            ]
          }
        },
        isInvalid: {
          // TODO: Not sure if needed
          displayName: "Invalid",
          type: "boolean",
          description: "Whether the input value is invalid",
          defaultValueHint: false
        },
        customValidationErrors: {
          // TODO: Not sure if needed
          type: "array",
          description: "Errors for custom validation"
        }
      }),
      states: {
        value: {
          type: "writable",
          valueProp: "value",
          onChangeProp: "onChange",
          variableType: "text"
        }
      },
      trapsFocus: true
    },
    overrides
  );
}

exports.BaseTextField = BaseTextField;
exports.TEXT_FIELD_COMPONENT_NAME = TEXT_FIELD_COMPONENT_NAME;
exports.registerTextField = registerTextField;
